beacon-chain/rpc/eth/helpers/sync.go

```diff
@@ -53,26 +55,134 @@ func ValidateSync(
+func IsOptimistic(
+	ctx context.Context,
+	stateId []byte,
+	optimisticModeFetcher blockchain.OptimisticModeFetcher,
+	stateFetcher statefetcher.Fetcher,
+	chainInfo blockchain.ChainInfoFetcher,
+	database db.ReadOnlyDatabase,
+) (bool, error) {
+	stateIdString := strings.ToLower(string(stateId))
+	switch stateIdString {
+	case "head":
+		return optimisticModeFetcher.IsOptimistic(ctx)
+	case "genesis", "finalized":
+		return false, nil
+	case "justified":
+		return isStateRootOptimistic(ctx, stateId, optimisticModeFetcher, stateFetcher, chainInfo)
+	default:
+		if len(stateId) == 32 {
+			return isStateRootOptimistic(ctx, stateId, optimisticModeFetcher, stateFetcher, chainInfo)
+		} else {
+			slotNumber, parseErr := strconv.ParseUint(stateIdString, 10, 64)
+			if parseErr != nil {
+				// ID format does not match any valid options.
+				e := statefetcher.NewStateIdParseError(parseErr)
+				return false, &e
+			}
+			optimistic, err := optimisticModeFetcher.IsOptimistic(ctx)
+			if err != nil {
+				return false, errors.Wrap(err, "could not check optimistic status")
+			}
+			if !optimistic {
+				return false, nil
+			}
+			headSt, err := stateFetcher.State(ctx, []byte("head"))
+			if err != nil {
+				return false, errors.Wrap(err, "could not fetch head state")
+			}
+			if primitives.Slot(slotNumber) == headSt.Slot() {
+				return optimisticModeFetcher.IsOptimistic(ctx)
+			}
+			finalizedSlot, err := slots.EpochStart(headSt.FinalizedCheckpoint().Epoch)
+			if err != nil {
+				return false, errors.Wrap(err, "could not get head state's finalized slot")
+			}
+			if primitives.Slot(slotNumber) <= finalizedSlot {
+				return false, nil
+			}
+			_, roots, err := database.BlockRootsBySlot(ctx, primitives.Slot(slotNumber))
+			if err != nil {
+				return false, errors.Wrapf(err, "could not get block roots for slot %d", slotNumber)
+			}
+			if len(roots) == 0 {
+				// TODO: What to do here?
+				return false, errors.New("no block")
+			}
+			for _, r := range roots {
+				canonical, err := chainInfo.IsCanonical(ctx, r)
+				if err != nil {
+					return false, errors.Wrapf(err, "could not check canonical status")
+				}
+				if canonical {
+					return optimisticModeFetcher.IsOptimisticForRoot(ctx, r)
+				}
+			}
+			// No block is canonical, return true
+			return true, nil
 		}
-		header.StateRoot = root[:]
 	}
-	headRoot, err := header.HashTreeRoot()
+}
+
+func isStateRootOptimistic(
+	ctx context.Context,
+	stateId []byte,
+	optimisticModeFetcher blockchain.OptimisticModeFetcher,
+	stateFetcher statefetcher.Fetcher,
+	chainInfo blockchain.ChainInfoFetcher,
+) (bool, error) {
+	st, err := stateFetcher.State(ctx, stateId)
+	if err != nil {
+		return false, errors.Wrap(err, "could not fetch justified state")
+	}
+	headSt, err := stateFetcher.State(ctx, []byte("head"))
+	if err != nil {
+		return false, errors.Wrap(err, "could not fetch head state")
+	}
+	if st.Slot() == headSt.Slot() {
+		return optimisticModeFetcher.IsOptimistic(ctx)
+	}
```

Comment:
you are grabbing a state only to check that the incoming id has the same state slot. You can use if you want `chainInfo.HeadSlot()` but also you should move this check (which is very unlikely) down the path.